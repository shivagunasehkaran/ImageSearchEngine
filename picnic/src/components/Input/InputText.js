import React from 'react';
import {TextInput, View, Image, Text, TouchableOpacity} from 'react-native';
import {styles} from '@components/Input/InputText.style';
import * as images from '@assets/styles/Images';

const InputText = props => {
  const {
    onChangeHandler,
    onFocus,
    onSubmit,
    onClear,
    isCancelButtonShown,
    handleCancel,
  } = props;
  return (
    <View style={styles.container}>
      <View style={styles.inputContainer}>
        <Image source={images.images.search} style={styles.searchIcon} />
        <TextInput
          testID={'searchTextInput'}
          style={styles}
          placeholder={'Search'}
          value={props.value}
          onChangeText={val => onChangeHandler(val)}
          onFocus={() => onFocus()}
          onSubmitEditing={() => onSubmit()}
          keyboardType={'email-address'}
          returnKeyType={'search'}
        />
        {isCancelButtonShown && (
          <TouchableOpacity style={styles.clearIcon} onPress={() => onClear()}>
            <Image source={images.images.clear} />
          </TouchableOpacity>
        )}
      </View>
      {isCancelButtonShown && (
        <TouchableOpacity style={styles.cancel} onPress={() => handleCancel()}>
          <Text>{'Cancel'}</Text>
        </TouchableOpacity>
      )}
    </View>
  );
};

export default InputText;
