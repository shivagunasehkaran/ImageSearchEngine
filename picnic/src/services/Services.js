import {GET_URL} from './Constants';

export async function getRandomAPI() {
  const url = GET_URL.RANDOM_URL;
  return fetch(url)
    .then(response => response.json())
    .then(responseJson => {
      return responseJson.data;
    })
    .catch(err => {
      return Promise.reject(err);
    });
}

export async function getSearchAPI(keyword) {
  const url = `${GET_URL.SEARCH_URL}&q=${keyword}&limit=9`;
  return fetch(url)
    .then(response => response.json())
    .then(responseJson => {
      return responseJson.data;
    })
    .catch(err => {
      return Promise.reject(err);
    });
}
