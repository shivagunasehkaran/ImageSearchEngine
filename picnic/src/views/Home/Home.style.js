import {ColourPalette} from '@assets/styles/ColoursStyles';
import {StyleSheet} from 'react-native';
import {moderateScale} from 'react-native-size-matters';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: ColourPalette.white,
  },
  stateText: {
    marginLeft: 20,
  },
  titleText: {
    marginHorizontal: 20,
  },
  renderItem: {
    flex: 1,
    height: moderateScale(100),
    margin: 5,
  },
});
