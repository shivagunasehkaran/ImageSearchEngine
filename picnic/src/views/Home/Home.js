import Details from '@components/Details';
import Error from '@components/Error';
import FastImage from '@components/FastImage';
import InputText from '@components/Input';
import {getRandomAPI, getSearchAPI} from '@services/Services';
import {styles} from '@views/Home/Home.style';
import React, {useEffect, useRef, useState} from 'react';
import {
  FlatList,
  Keyboard,
  SafeAreaView,
  Text,
  TouchableOpacity,
  View,
  ActivityIndicator,
} from 'react-native';
import * as ROUTES from '@routes/Routes';
import {ConstantsText} from '@utills/Constants';

const Home = props => {
  const [keyword, setKeyword] = useState('');
  const [stateName, setStateName] = useState(ConstantsText.randomSelected);
  const [keyBoardFocus, setKeyBoardFocus] = useState(false);
  const [randomData, setRandomData] = useState([]);
  const [searchData, setSearchData] = useState([]);
  const [showCancel, setShowCancel] = useState(false);
  const [isLoading, setLoading] = useState(false);
  const interval = useRef();

  // Remember the latest callback.
  useEffect(() => {
    setLoading(true);
    interval.current = setInterval(() => callRandomAPI(), 10000);
    return () => {
      clearInterval(interval.current);
    };
  }, []);

  // random API call
  async function callRandomAPI() {
    let data = await getRandomAPI();
    setRandomData(data);
    setLoading(false);
  }

  // search API calls after typed 2 chars
  const handleSearchInput = async val => {
    setKeyword(val);
    if (keyword.length == 1) {
      let data = await getSearchAPI(keyword);
      // set search data
      setSearchData(data);
    }
  };

  // go to image details screen
  const goToDetail = item => {
    props.navigation.push(ROUTES.pageNameImageDetails, {item: item});
  };

  // flat list render item
  const renderRow = items => {
    const item = items.item;
    let URL = null;
    if (item && item.images && item.images.downsized_large) {
      URL = item.images.downsized_large.url;
    }
    return (
      <View style={styles.renderItem}>
        <TouchableOpacity
          onPress={() => goToDetail(item)}
          testID={'itemAction'}>
          <FastImage uri={URL} />
        </TouchableOpacity>
      </View>
    );
  };

  // on focus listener
  const onFocusHandler = () => {
    setKeyBoardFocus(true);
    setShowCancel(true);
    setStateName(ConstantsText.searchResults);
    clearInterval(interval.current); // Clear the interval here
  };

  const handleCancel = () => {
    setKeyword('');
    setShowCancel(false);
    setKeyBoardFocus(false);
    setStateName(ConstantsText.randomSelected);
    Keyboard.dismiss();
    interval.current = setInterval(() => callRandomAPI(), 10000);
  };

  const handleClear = () => {
    setKeyword('');
  };

  // calling api for on submit (keyboard)
  const onSubmitHandler = () => {
    Keyboard.dismiss();
  };

  return (
    <View style={styles.container}>
      <SafeAreaView>
        <InputText
          value={keyword}
          onChangeHandler={val => handleSearchInput(val)}
          onFocus={() => onFocusHandler()}
          showCancel={showCancel}
          handleCancel={() => handleCancel()}
          onClear={() => handleClear()}
          onSubmit={() => onSubmitHandler()}
        />
        <Text style={styles.stateText}>{stateName}</Text>
        {isLoading && <ActivityIndicator size="large" />}
        {keyBoardFocus ? (
          <View style={{margin: 20}}>
            <FlatList
              data={searchData}
              renderItem={renderRow}
              keyExtractor={item => item.id.toString()}
              numColumns={3}
            />
          </View>
        ) : (
          <Details randomData={randomData} />
        )}
      </SafeAreaView>
      {/* {!isLoading && <>{!randomData.length ? <Error /> : null}</>}   Error component created to validate { minimal API calls {API rate exceeded} */}
    </View>
  );
};

export default Home;
