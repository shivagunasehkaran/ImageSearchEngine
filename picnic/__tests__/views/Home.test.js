import * as ServicesAPI from '@services/Services';
import Home from '@views/Home';
import enzyme, {mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import {Alert} from 'react-native';
import renderer from 'react-test-renderer';
import InputText from '@components/Input/InputText';

jest.mock('@services/Services');

enzyme.configure({adapter: new Adapter()});

describe('Home screen', () => {
  it('renders correctly with all home screen values', async () => {
    const wrapper = shallow(<Home />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render input element properly', async () => {
    const wrapper = shallow(<InputText value={'keyword'} />);
    expect(wrapper).toMatchSnapshot();
  });

  // integration testing
  it('render search results when keyword changes', async () => {
    const onSearch = jest.fn();
    const wrapper = shallow(<InputText value={'keyword'} />);
    wrapper.find('TextInput').simulate('changeTextHandler', 'test search text');
    expect(onSearch).toHaveBeenCalledWith('test search text');
  });

  it('should render the home screen without breaking', async () => {
    const rendered = renderer.create(<Home />).toJSON();
    expect(rendered).toBeTruthy();
  });

  it('should render state text element', async () => {
    const wrapper = shallow(<Home />);
    expect(wrapper.find('Text')).toHaveLength(1);
  });

  it('should flatlist return keyExtractor correctly', () => {
    const item = [
      {
        id: '1',
        age: 30,
        name: 'shiva',
      },
      {
        id: '2',
        age: 31,
        name: 'gautham',
      },
    ];
    const wrapper = shallow(<Home item={item} />);
    expect(wrapper.find('FlatList').length).toEqual(0);
  });

  it('navigate: renders correctly with default', async () => {
    const fakeNavigation = {
      push: jest.fn(),
    };
    const wrapper = mount(<Home navigation={fakeNavigation} />);
    var home = wrapper.find('Home');
    const node = home
      .findWhere(node => node.prop('testID') === 'itemAction')
      .at(0);
    node.simulate('click');
  });
});

describe('Home screen : random API', () => {
  it('random api success', async () => {
    const spy = jest
      .spyOn(ServicesAPI, 'getRandomAPI')
      .mockImplementation(url => {
        return new Promise(resolve => resolve('test'));
      });
    await new Promise(resolve => setTimeout(resolve, 100));
    expect(spy).toHaveBeenCalledWith({url: 'http://dummyURL.com/'});
  });

  it('random api with error message', async () => {
    const alertSpy = jest.spyOn(Alert, 'alert');
    jest.spyOn(ServicesAPI, 'getRandomAPI').mockImplementation(url => {
      return new Promise(resolve => resolve([]));
    });
    await new Promise(resolve => setTimeout(resolve, 100));
    expect(alertSpy).toBeCalled();
    wrapper.update();
    alertSpy.mock.calls[1][2][0].onPress();
  });
});

describe('Home screen : search API', () => {
  it('search api success', async () => {
    const spy = jest
      .spyOn(ServicesAPI, 'getSearchAPI')
      .mockImplementation(url => {
        return new Promise(resolve => resolve(['test']));
      });

    const wrapper = shallow(<Home />);
    var homeView = wrapper.find('Home');
    const node = homeView
      .findWhere(node => node.prop('testID') === 'itemAction')
      .at(0);
    node.simulate('click');
    expect(spy).toHaveBeenCalledWith({url: 'http://dummyURL.com/'});
  });

  it('search api error', async () => {
    jest.spyOn(ServicesAPI, 'getSearchAPI').mockImplementation(() => {
      return new Promise(resolve => resolve([]));
    });
    const wrapper = shallow(<Home />);
    var homeView = wrapper.find('Home');
    const node = homeView
      .findWhere(node => node.prop('testID') === 'itemAction')
      .at(0);
    node.simulate('click');
    expect(node).toEqual({});
  });
});
