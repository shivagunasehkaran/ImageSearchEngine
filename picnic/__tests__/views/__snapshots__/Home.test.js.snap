// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen renders correctly with all home screen values 1`] = `ShallowWrapper {}`;

exports[`Home screen should render input element properly 1`] = `ShallowWrapper {}`;
